### PR TITLE
ddl: fix the 'max-index-length' check result in non-restricted sql mod 

### LINKED ISSUE #34931: ddl: the 'max-index-length' check does not respect non-restricted sql_mode
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
drop table if exists t;
set @@sql_mode='';
create table t (id int, name varchar(2048), index(name)) charset=utf8;
```

### 2. What did you expect to see? (Required)
In MySQL 5.7 & MySQL 8:

```
mysql> set @@sql_mode='';
Query OK, 0 rows affected (0.00 sec)

mysql> create table t (id int, name varchar(2048), index(name)) charset=utf8;
Query OK, 0 rows affected, 2 warnings (0.02 sec)

mysql> show warnings;
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                                                     |
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 3719 | 'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous. |
| Warning | 1071 | Specified key was too long; max key length is 3072 bytes                                                                                                                    |
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set (0.00 sec)

mysql> show create table t;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                          |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `id` int(11) DEFAULT NULL,
  `name` varchar(2048) DEFAULT NULL,
  KEY `name` (`name`(1024))
) ENGINE=InnoDB DEFAULT CHARSET=utf8 |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

Notice that in non-strict SQL mode, MySQL converted the error to warning and truncated the index `name` as a prefix one("`name`(1024)"), which met the maximum length of index(3072 bytes / 3 bytes per utf8 encoding = 1024).

### 3. What did you see instead (Required)
In TiDB(all versions with default configuration):

```
mysql> set @@sql_mode='';
Query OK, 0 rows affected (0.00 sec)

tidb> create table t (id int, name varchar(2048), index(name)) charset=utf8;
ERROR 1071 (42000): Specified key was too long; max key length is 3072 bytes
```

### 4. What is your TiDB version? (Required)

```
tidb> select version();
+--------------------+
| version()          |
+--------------------+
| 5.7.25-TiDB-v6.0.0 |
+--------------------+
1 row in set (0.00 sec)
```

I am using v6.0 but it applies to all existing versions.

### PR DESCRIPTION
**Pull request by e11jah**
Signed-off-by: elijah <f1renze.142857@gmail.com>

### What problem does this PR solve?

Issue Number: close #34931

Problem Summary: the 'max-index-length' check does not respect non-restricted sql_mode

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
```sql
mysql> set @@sql_mode='';
Query OK, 0 rows affected (0.00 sec)

mysql> create table t (id int, name varchar(2048), index(name)) charset=utf8;
Query OK, 0 rows affected, 2 warnings (0.02 sec)

mysql> show warnings;
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                                                     |
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 3719 | 'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous. |
| Warning | 1071 | Specified key was too long; max key length is 3072 bytes                                                                                                                    |
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set (0.00 sec)

mysql> show create table t;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                          |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `id` int(11) DEFAULT NULL,
  `name` varchar(2048) DEFAULT NULL,
  KEY `name` (`name`(1024))
) ENGINE=InnoDB DEFAULT CHARSET=utf8 |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
converted the error to warning in non-strict SQL mode with `max-index-length` check
```

### CODE REVIEW

**Comment by bb7133** [comment]
@e11jah , just a kind reminder:

The `max-index-length' in MySQL(with InnoDB) is [hard-coded to 3072](https://dev.mysql.com/doc/refman/8.0/en/innodb-limits.html#:~:text=The%20index%20key%20prefix%20length,REDUNDANT%20or%20COMPACT%20row%20format.), but it is configurable in TiDB because of some histroical reason:

https://docs.pingcap.com/tidb/dev/tidb-configuration-file#max-index-length

File: ddl/db_test.go [hunk 1]
```diff
@@ -1470,3 +1470,24 @@ func TestReportingMinStartTimestamp(t *testing.T) {
...
+	_, err := tk.Exec("create table t1 (id int, name varchar(2048), index(name)) charset=utf8;")
+	require.Error(t, err)
+	tk.MustExec("set @@sql_mode=''")
+
+	_, err = tk.Exec("create table t2 (id int, name varchar(2048), index(name)) charset=utf8;")
+	require.NoError(t, err)
+	require.Equal(t, uint16(1), tk.Session().GetSessionVars().StmtCtx.WarningCount())
+
+	_, err = tk.Exec("create table t1 (id int, name varchar(2048), unique index(name)) charset=utf8;")
+	require.Error(t, err)
```
[the remark this answers is not in the document]
Comment (e11jah) [reply ddl/db_test.go:1492]: got it

**Review by CbcWestwolf — APPROVED** [review]

File: ddl/db_test.go [hunk 2]
```diff
@@ -1470,3 +1470,40 @@ func TestReportingMinStartTimestamp(t *testing.T) {
...
+	sql := "create table t1 (id int, name varchar(2048), index(name)) charset=utf8;"
+	tk.MustGetErrCode(sql, errno.ErrTooLongKey)
+
+	r := tk.MustQuery("select @@sql_mode")
+	defer func(sqlMode string) {
+		tk.MustExec("set @@sql_mode= '" + sqlMode + "'")
+		tk.MustExec("drop table if exists t1")
+		tk.MustExec("drop table if exists t2")
+		tk.MustExec("drop table if exists t3")
+	}(r.Rows()[0][0].(string))
```
Comment (tangenta) [inline ddl/db_test.go:1496]:
```suggestion
```

We don't have to set it back because each test is isolated.

File: ddl/index.go [hunk 3]
```diff
@@ -78,13 +78,22 @@ func buildIndexColumns(columns []*model.ColumnInfo, indexPartSpecifications []*a
...
+			if ctx == nil || ctx.GetSessionVars().StrictSQLMode || mysql.HasUniKeyFlag(col.GetFlag()) || len(indexPartSpecifications) > 1 {
+				return nil, dbterror.ErrTooLongKey.GenWithStackByArgs(config.GetGlobalConfig().MaxIndexLength)
+			}
+			colLenPerUint, err := getIndexColumnLength(col, 1)
+			if err != nil {
+				return nil, err
+			}
+			indexColLen = config.GetGlobalConfig().MaxIndexLength / colLenPerUint
+			// produce warning message
+			ctx.GetSessionVars().StmtCtx.AppendWarning(dbterror.ErrTooLongKey.FastGenByArgs(config.GetGlobalConfig().MaxIndexLength))
```
Comment (tangenta) [inline ddl/index.go:90]: `config.GetGlobalConfig().MaxIndexLength` occurs multiple times. Please use a local variable to represent it.

**Review by tangenta — APPROVED** [review]

File: ddl/db_test.go [hunk 4]
```diff
@@ -1470,3 +1470,38 @@ func TestReportingMinStartTimestamp(t *testing.T) {
...
+	// For a unique index, an error occurs regardless of SQL mode because reducing
+	//the index length might enable insertion of nonunique entries that do not meet
+	//the specified uniqueness requirement.
+	sql = "create table t1 (id int, name varchar(2048), unique index(name)) charset=utf8;"
+	tk.MustGetErrCode(sql, errno.ErrTooLongKey)
+
+	// The multiple column index in which the length sum exceeds the maximum size
+	// will return an error instead produce a warning.
+	sql = "create table t3 (id int, name varchar(512), alias varchar(1024), index(name,alias)) charset=utf8;"
+	tk.MustGetErrCode(sql, errno.ErrTooLongKey)
```
Comment (bb7133) [inline ddl/db_test.go:1506]: Please add the test case for more charsets:

* `ascii`, for which `colLenPerUint` is 1
* `utf8mb4`, for which `colLenPerUint` is 4

Also, please verify the result for `char`, which might(I'm not sure) be different with `varchar`.

Thank you!

**Comment by e11jah** [comment]
> Is there any reason this PR is changed back to 'WIP'?

Sorry, I find that there is a problem with text type column, I will push the commit later.

**Comment by bb7133** [comment]
> Sorry, I found that the error occurs 

@e11jah You don't need to say 'sorry' :)

File: ddl/db_test.go [hunk 5]
```diff
@@ -1470,3 +1474,120 @@ func TestReportingMinStartTimestamp(t *testing.T) {
...
+// for issue #34931
+func TestBuildMaxLengthIndexWithNonRestrictedSqlMode(t *testing.T) {
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+
+	r := tk.MustQuery("select @@sql_mode")
+	defaultSqlMode := r.Rows()[0][0].(string)
```
Comment (bb7133) [inline ddl/db_test.go:1487]:
```suggestion
	defaultSQLMode := r.Rows()[0][0].(string)
```
From https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_check/detail/tidb_ghpr_check/65639/pipeline:

```
[2022-06-01T12:24:31.444Z] linting
[2022-06-01T12:24:31.444Z]   ✘  https://revive.run/r#var-naming  var defaultSqlMode should be defaultSQLMode  
[2022-06-01T12:24:31.444Z]   ddl/db_test.go:1487:2
```

Comment (e11jah) [reply ddl/db_test.go]: fixed

File: ddl/index.go [hunk 6]
```diff
@@ -66,25 +66,37 @@ func buildIndexColumns(columns []*model.ColumnInfo, indexPartSpecifications []*a
...
 			return nil, err
 		}
 		sumLength += indexColumnLength
 
 		// The sum of all lengths must be shorter than the max length for prefix.
-		if sumLength > config.GetGlobalConfig().MaxIndexLength {
-			return nil, dbterror.ErrTooLongKey.GenWithStackByArgs(config.GetGlobalConfig().MaxIndexLength)
+		if sumLength > maxIndexLength {
+			// The multiple column index and the unique index in which the length sum exceeds the maximum size
+			// will return an error instead produce a warning.
```
Comment (bb7133) [inline ddl/index.go:82]:
```suggestion
			// will return an error instead produce a warning in strict sql mode.
```
How about making it more clear?

File: ddl/db_test.go [hunk 7]
```diff
@@ -1470,3 +1474,123 @@ func TestReportingMinStartTimestamp(t *testing.T) {
...
+	}
+
+	sqlTemplate := "create table %s (id int, name %s, age int, %s index(name%s%s)) charset=%s;"
+	// test character strings for varchar and text
+	for _, tc := range tt {
+
+		for _, cs := range charset.CharacterSetInfos {
+			tableName := fmt.Sprintf("t_%s", cs.Name)
+			tk.MustExec(fmt.Sprintf("drop table if exists %s", tableName))
+			tk.MustExec(fmt.Sprintf("set @@sql_mode='%s'", defaultSQLMode))
```
Comment (bb7133) [inline ddl/db_test.go:1525]: You could simply use "`set @@sql_mode=default`" to set it to default.

**Comment by bb7133** [comment]
The rest LGTM.

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**